### ISSUE TITLE
Is there any way to lazily create the DropdownMenu items?

### ISSUE THREAD

**Issue by gaearon** [nandorojo/zeego#114]
I realize this may be a request to `react-native-menu`, but I'll file here.

Currently we specify dropdown menu content by putting stuff into `<DropdownMenu.Content>`. The problem is that we have to put it there _eagerly_. In other words, we have to prepare `<DropdownMenu.Item>`s in the parent component _in case_ the menu gets opened. The downside is that their event handlers might need certain Hooks (e.g. RQ `useMutation`) so we end up bloating the parent component render time even if the menu never gets opened.

Ideally I'd like to be able to split the content so that it's lazily initialized, but there doesn't seem to be a way to do that. I was hoping to do something like rendering it conditionally based on state driven by `<DropdownMenuRoot onOpenWillChange>` but that doesn't seem possible either (the event simply doesn't exist on Android).

Is there any way to achieve what I want?

**Comment by nandorojo** [comment]
@gaearon hey sorry I just saw this. I like the idea of inflating the menu items lazily. Let me see if there is a similar callback for Android to implement something like this.

**Comment by nandorojo** [comment]
As you mentioned though, for the Android side we’d need a callback for react-native-menu (assuming we stick with that library). I’ll see if that’s an option.

**Comment by nandorojo** [comment]
@gaearon FYI - I am working on this but it requires a full rewrite of Zeego.

TLDR: the upstream native libs aren't iterating any time soon so I'm rewriting the native elements directly in Zeego using SwiftUI and Jetpack Compose. This is coming in Zeego 3.

As long as you're using the new arch you'll be able to use Zeego 3. 

By controlling the native code we won't need to wait forever for the upstream libs to add something as simple as a callback. As long as Android supports it we can add it.

I'm currently pretty far along with the iOS rewrite in SwiftUI. Other than `ContextMenu.Auxiliary` (used for iMessage-style reactions) it should be easy to port everything to native.

Still TBD on Android timeline but I imagine it won't be that hard either.

**Comment by nandorojo** [comment]
As an update, the entire rewrite in SwiftUI is complete. **However, there are 2 hold ups:**

1. **Callbacks (what OP wanted):** SwiftUI provides no callbacks for when a menu opens/closes. The best you can do is add `.onAppear` to the first child inside the menu. However...this only fires the first time. SwiftUI strangely does _not_ "unmount" menu items after they close, meaning they never fire `.onDisappear`. So it's unclear how to solve this for now. If anyone knows, I'd love a pointer.
2. `<ContextMenu.Auxiliary>` Similarly, I don't know how to support this in SwiftUI at the moment, as hinted at in #133. `react-native-ios-context-menu` does a lot of magic to get this working via [ContextMenuAuxiliaryPreview](https://github.com/dominicstop/ContextMenuAuxiliaryPreview). I'm not entirely sure how to integrate that library into Zeego.

To proceed: I'm going to do my best to get 1) callbacks working, and as for 2...TBD. Maybe it'll be a versioned thing or there can be a `zeego/new` for the time being until I can assess how many people use that iOS-only component.

**Comment by nandorojo** [comment]
# Workaround

For anyone who wants this, my best workaround might be to add some sort of touch/long press listeners to the child of `DropdownMenu.Trigger`

**Comment by nandorojo** [comment]
Hey guys, Zeego 3 is out. As of `3.0.3`, all platforms support `onOpenChange`, so this works now.

- [X Post](https://x.com/FernandoTheRojo/status/1897380273378979854)
- [Release Notes](https://t.co/xwuo8YIM3c)
